### ISSUE TITLE
The problem of Mamba block input consistency but output fluctuation in model validation environment

### ISSUE THREAD

**Issue by xypjq** [state-spaces/mamba#548]
1. In my task, I used a nested network structure with multiple layers of MAMBA blocks, but in model validation mode, I found that when I gave the same model input, there were different model outputs, and some of these outputs had significant fluctuations. At first, I suspected that it might be a problem with my other modules, but in the end, I was able to pinpoint that it was caused by the internal operations of the MAMBA block. It could be a loss of accuracy. How can I try to avoid this problem and hope to receive relevant answers.
2. Attached below is a diagram of the problem situation：
![image](https://github.com/user-attachments/assets/c8720c4d-bd8f-4c35-9b8c-10a0600ed2ac)age.png

**Comment by zz202** [comment]
@xypjq Were you able to find a solution to this? I noticed the similar issue, reproducing previous work without changing any settings, using the same input... however, got inconsistant outputs, and some of the outputs showed significant fluctuations, as you commented.

**Comment by Liujehong** [comment]
@xypjq @zz202 I encountered the same issue in my experiments. Due to the influence of other modules, the errors caused by the Mamba module are amplified! Have you resolved this problem?

**Comment by Liujehong** [comment]
@zz202 @xypjq  According to the response in https://github.com/state-spaces/mamba/issues/493, the implementation of Mamba includes non-deterministic atomic operations (atomic add). So, I suspect you and I are encountering the same issue, where the output inconsistency occurs during the validation phase after training. In fact, it's not just the output that is inconsistent—if you observe the network parameters, they will also be slightly different. Unfortunately, this is an issue in PyTorch that cannot be resolved at the moment.

**Comment by xypjq** [comment]
@Liujehong @zz202 In my code, I identified the issue as occurring in Mamba's own implementation of layernorm RMSnorm. My solution is to replace Mamba's layernorm with Torch's implementation.I hope it can help you.
